### ISSUE TITLE
Shared examples context uses NSDictionary, but it probably shouldn't

### ISSUE THREAD

**Issue by modocache** [Quick/Quick#153]
Parameters passed in via a shared examples context should not have to be Objective-C objects. That is, I should be able to do the following:

``` swift
sharedExamples("shared examples that take a context") { (sharedExampleContext: SharedExampleContext) in
  it("is passed the correct parameters via the context") {
    let wasPassedIn = sharedExampleContext()["wasPassedIn"] as Boolean // currently error: Type 'Boolean' does not conform to protocol 'AnyObject'
    expect(wasPassedIn).to(beTruthy())
  }
}
```

**Comment by nikita-leonov** [comment]
This is what I mean about using tuples in shared examples — http://swiftstub.com/561242800/
Such sharedExamples implementation allows to avoid typecasting boilerplate in function that happens all the time. 
Disregard world implementation, It just for the demo purposes. Also I do know that generics in such case is not the best option, but I didn't find a better way to work with not pre-defined tuples.

**Comment by abbeycode** [comment]
Am I missing something, or is a `[String: Any]` the right (and simplest) data structure here? I just tested in an Xcode 6 playground that anonymous tuples, structs, and classes can all live together. You'll still need to cast to the type you're expecting when you extract data, which is the same as with NSDictionary.

We could also go a step beyond and make it a generic wrapper that's backed by [String: T], so that if you want to use it in a strongly typed manner, you could.

**Comment by modocache** [comment]
@abbeycode That all sounds great! Would that data structure be usable from Objective-C?

**Comment by abbeycode** [comment]
No, using a `Swift.Dictionary` would make it Swift-only. It would be an alternative, not a replacement.

**Comment by nikita-leonov** [comment]
How do you expect [String: T] to work? Every key in shared example may have different type. Also for a [String: Any] this is what I tried to avoid with tuples, as you can condense type casting into shared example definition. Check the link above for a possible solution. It would not work so elegant with Any in Dictionary.

**Comment by abbeycode** [comment]
@nikita-leonov Ah, I had skimmed over that before. That's a cool way to handle it. The only issue I see with that approach is that the keys in that case don't really matter - only their order does. They can mismatch between the `itBehavesLike` invocation and the `sharedExamples` definition, and everything will work, so long as the order of the tuple values matches up.

That's not necessarily bad, but you're trading an indexed key for the string key, and getting type safety for all values in return. There's just a chance that you might inadvertently rearrange two of the tuple's members in either place, and it would look good (if their types match) but then produce incorrect results. Overall, that sounds like a good solution to me, though.

**Comment by jeffh** [comment]
:+1: for the shared tuple. I don't mind the ordering trade-off.

**Comment by briancroom** [comment]
I was thinking earlier about how to make shared examples Swiftier and had exactly the same idea as @nikita-leonov demo'd on SwiftStub. I was about to open an issue proposing this addition but found this one first. 

The beautiful thing about this approach is that you are still free to use a dictionary as the context type if you feel it makes send for your tests, but it's also possible to move away from stringly-typed data if you can model it better. 

One particularly nice use case is when using a shared example as a "contract test" (hat tip @camelpunch), you can define the context type to be just the protocol whose contract you are specifying. No need to wrap the object in a dictionary with an arbitrary key.

**Comment by istx25** [comment]
Has there been any progress on this?

**Comment by ikesyo** [comment]
Looks like this has been addressed by #654.

**Comment by istx25** [comment]
Ah sweet. Sounds good. Closing…
